File: frontend/src/features/appReview/api/mapper/appReview.mapper.ts
import type {
  AppReviewResponseDto,
  AppReviewSummaryResponseDto,
  PublicAppReviewResponseDto,
} from '@/features/appReview/api/dtos/response.dto';
import type {
  AppReview,
  AppReviewSummary,
  PublicAppReview,
} from '@/features/appReview/types/contracts';

export const mapAppReviewResponseDto = (payload: AppReviewResponseDto): AppReview => {
  return {
    id: payload.id,
    userId: payload.userId,
    displayName: payload.displayName,
    rating: payload.rating,
    comment: payload.comment,
    isHidden: payload.isHidden,
    createdAt: payload.createdAt,
    updatedAt: payload.updatedAt,
  };
};

export const mapPublicAppReviewResponseDto = (payload: PublicAppReviewResponseDto): PublicAppReview => {
  return {
    id: payload.id,
    displayName: payload.displayName,
    rating: payload.rating,
    comment: payload.comment,
    createdAt: payload.createdAt,
    updatedAt: payload.updatedAt,
  };
};

export const mapAppReviewSummaryResponseDto = (payload: AppReviewSummaryResponseDto): AppReviewSummary => {
  return {
    averageRating: Number(payload.averageRating ?? 0),
    totalReviews: Number(payload.totalReviews ?? 0),
    fiveStarCount: Number(payload.fiveStarCount ?? 0),
    fourStarCount: Number(payload.fourStarCount ?? 0),
    threeStarCount: Number(payload.threeStarCount ?? 0),
    twoStarCount: Number(payload.twoStarCount ?? 0),
    oneStarCount: Number(payload.oneStarCount ?? 0),
  };
};

File: backend/fookbase.API/Src/Application/Mappers/System/AppReviewMapper.cs
using InteractHub.Api.Application.DTOs.AppReviews;
using InteractHub.Api.Domain.Entities;

namespace InteractHub.Api.Application.Mappers;

public static class AppReviewMapper
{
    public static AppReviewSummaryResponseDto ToSummaryResponseDto(
        IReadOnlyDictionary<int, int> distribution,
        double average,
        int totalCount)
    {
        ArgumentNullException.ThrowIfNull(distribution);

        return new AppReviewSummaryResponseDto
        {
            AverageRating = Math.Round(average, 1, MidpointRounding.AwayFromZero),
            TotalReviews = totalCount,
            FiveStarCount = distribution.TryGetValue(5, out var fiveStarCount) ? fiveStarCount : 0,
            FourStarCount = distribution.TryGetValue(4, out var fourStarCount) ? fourStarCount : 0,
            ThreeStarCount = distribution.TryGetValue(3, out var threeStarCount) ? threeStarCount : 0,
            TwoStarCount = distribution.TryGetValue(2, out var twoStarCount) ? twoStarCount : 0,
            OneStarCount = distribution.TryGetValue(1, out var oneStarCount) ? oneStarCount : 0
        };
    }

    public static AppReviewResponseDto ToResponseDto(this AppReview review)
    {
        ArgumentNullException.ThrowIfNull(review);

        return new AppReviewResponseDto
        {
            Id = review.Id,
            UserId = review.UserId,
            DisplayName = review.DisplayName,
            Rating = review.Rating,
            Comment = review.Comment,
            IsHidden = review.IsHidden,
            CreatedAt = review.CreatedAt,
            UpdatedAt = review.UpdatedAt
        };
    }

    public static PublicAppReviewResponseDto ToPublicResponseDto(this AppReview review)
    {
        ArgumentNullException.ThrowIfNull(review);

        return new PublicAppReviewResponseDto
        {
            Id = review.Id,
            DisplayName = review.DisplayName,
            Rating = review.Rating,
            Comment = review.Comment,
            CreatedAt = review.CreatedAt,
            UpdatedAt = review.UpdatedAt
        };
    }
}



